go/pkg/client/cas.go

```diff
@@ -25,32 +26,131 @@ import (
 	log "github.com/golang/glog"
 )
 
-// UploadIfMissing stores a number of uploadable items.
-// It first queries the CAS to see which items are missing and only uploads those that are.
-// Returns a slice of the missing digests.
-func (c *Client) UploadIfMissing(ctx context.Context, data ...*chunker.Chunker) ([]digest.Digest, error) {
-	if cap(c.casUploaders) <= 0 {
-		return nil, fmt.Errorf("CASConcurrency should be at least 1")
+type uploadRequest struct {
+	c    *chunker.Chunker
+	wait chan<- *uploadResponse
+}
+
+type uploadResponse struct {
+	digest  digest.Digest
+	err     error
+	missing bool
+}
+
+type uploadState struct {
+	data *chunker.Chunker
+	err  error
+
+	// mu protects clients. It does NOT protect data or error, because they do not need protection -
+	// they are only modified when a state object is created, and by updateAndNotify which is called
+	// exactly once for a given state object (this is the whole point of the algorithm).
+	mu      sync.Mutex
+	clients []chan<- *uploadResponse
+}
+
+func (c *Client) getMissingPresent(ctx context.Context, dgs []digest.Digest) (missing []digest.Digest, present []digest.Digest, err error) {
+	dgMap := make(map[digest.Digest]bool)
+	for _, d := range dgs {
+		dgMap[d] = true
+	}
+	missing, err = c.MissingBlobs(ctx, dgs)
+	for _, d := range missing {
+		delete(dgMap, d)
+	}
+	for d := range dgMap {
+		present = append(present, d)
+	}
+	return missing, present, err
+}
+
+func (c *Client) uploadProcessor() {
+	var buffer []*uploadRequest
```

Comment:
I don't fully understand why we need this buffer. It is already buffered via casUploadRequests channel? Can you call `upload()` as and when you see new items in `casUploadRequests` ?

Reply:
> If you only read from the channel, you are in danger of never stopping (if new items keep arriving)

I see, makes sense.